File: Apps/HIL/android/app/src/main/java/com/dweekly/cyrinxhil/DTMFCodec.kt
package com.dweekly.cyrinxhil

import kotlin.math.PI
import kotlin.math.cos
import kotlin.math.max
import kotlin.math.min
import kotlin.math.sin
import kotlin.math.sqrt

class DTMFCodec(config: SessionConfig) {
    private val sampleRateHz = config.sampleRateHz.coerceIn(8_000, 192_000)
    private val symbolSamples = config.dcssSymbolSamples.coerceIn(960, 8192)
    private val txGainCap = config.txGainCap.coerceIn(0.01f, 0.80f)
    private val leaderSymbols = 6
    private val gapSymbols = 2
    private val minLeaderRms = 0.0025f
    private val repetitionCount = 3
    private val syncNibbles = intArrayOf(0x1, 0x2, 0x3, 0x4, 0xB, 0xC, 0xD, 0xE)
    private val rowHz = intArrayOf(697, 770, 852, 941)
    private val colHz = intArrayOf(1209, 1336, 1477, 1633)
    private val rowRefs = rowHz.map { makeReference(it, txGainCap * 0.5f) }
    private val colRefs = colHz.map { makeReference(it, txGainCap * 0.5f) }
    private val rxBuffer = ArrayList<Float>()

    fun encode(payload: ByteArray): FloatArray {
        val trimmed = payload.copyOfRange(0, minOf(payload.size, 255))
        val crcInput = ByteArray(trimmed.size + 1)
        crcInput[0] = trimmed.size.toByte()
        System.arraycopy(trimmed, 0, crcInput, 1, trimmed.size)
        val crc = crc8(crcInput, 0, crcInput.size)

        val nibbles = ArrayList<Int>(syncNibbles.size + (4 * repetitionCount) + (trimmed.size * 2 * repetitionCount))
        for (nibble in syncNibbles) {
            nibbles.add(nibble)
        }
        repeat(repetitionCount) { nibbles.add((trimmed.size ushr 4) and 0xF) }
        repeat(repetitionCount) { nibbles.add(trimmed.size and 0xF) }
        for (byte in trimmed) {
            repeat(repetitionCount) { nibbles.add((byte.toInt() ushr 4) and 0xF) }
            repeat(repetitionCount) { nibbles.add(byte.toInt() and 0xF) }
        }
        repeat(repetitionCount) { nibbles.add((crc ushr 4) and 0xF) }
        repeat(repetitionCount) { nibbles.add(crc and 0xF) }

        val totalSymbols = leaderSymbols + gapSymbols + nibbles.size
        val out = FloatArray(totalSymbols * symbolSamples)
        var cursor = 0
        repeat(leaderSymbols) {
            val symbol = synthesizeSymbol(0)
            System.arraycopy(symbol, 0, out, cursor, symbolSamples)
            cursor += symbolSamples
        }
        cursor += gapSymbols * symbolSamples
        for (nibble in nibbles) {
            val symbol = synthesizeSymbol(nibble)
            System.arraycopy(symbol, 0, out, cursor, symbolSamples)
            cursor += symbolSamples
        }
        return out
    }

    fun ingest(samples: FloatArray): List<ByteArray> {
        if (samples.isEmpty()) {
            return emptyList()
        }
        for (sample in samples) {
            rxBuffer.add(sample)
        }
        return decodeAvailable()
    }

    private fun decodeAvailable(): List<ByteArray> {
        val decoded = ArrayList<ByteArray>()
        val minSymbols = leaderSymbols + gapSymbols + syncNibbles.size + (4 * repetitionCount)
        val minSamples = minSymbols * symbolSamples

        while (rxBuffer.size >= minSamples) {
            val leaderStart = findLeaderStart() ?: break
            val nominalStart = leaderStart + ((leaderSymbols + gapSymbols) * symbolSamples)
            val start = refineSyncStart(nominalStart)
            if (start == null) {
                dropFront(leaderStart + symbolSamples)
                continue
            }
            val lenHi = decodeRepeatedNibbleAt(start, syncNibbles.size)
            val lenLo = decodeRepeatedNibbleAt(start, syncNibbles.size + repetitionCount)
            if (lenHi == null || lenLo == null) {
                dropFront(leaderStart + symbolSamples)
                continue
            }
            val length = ((lenHi shl 4) or lenLo) and 0xFF
            val totalNibbles = syncNibbles.size + (2 * repetitionCount) + (length * 2 * repetitionCount) + (2 * repetitionCount)
            val frameEnd = start + (totalNibbles * symbolSamples)
            if (rxBuffer.size < frameEnd) {
                if (leaderStart > 0) {
                    dropFront(leaderStart)
                }
                break
            }

            val payload = ByteArray(length)
            var nibbleIndex = syncNibbles.size + (2 * repetitionCount)
            var valid = true
            for (i in 0 until length) {
                val hi = decodeRepeatedNibbleAt(start, nibbleIndex)
                val lo = decodeRepeatedNibbleAt(start, nibbleIndex + repetitionCount)
                if (hi == null || lo == null) {
                    valid = false
                    break
                }
                payload[i] = (((hi shl 4) or lo) and 0xFF).toByte()
                nibbleIndex += 2 * repetitionCount
            }
            if (!valid) {
                dropFront(leaderStart + symbolSamples)
                continue
            }

            val crcHi = decodeRepeatedNibbleAt(start, nibbleIndex)
            val crcLo = decodeRepeatedNibbleAt(start, nibbleIndex + repetitionCount)
            if (crcHi == null || crcLo == null) {
                dropFront(leaderStart + symbolSamples)
                continue
            }
            val actualCrc = ((crcHi shl 4) or crcLo) and 0xFF
            val crcInput = ByteArray(length + 1)
            crcInput[0] = length.toByte()
            System.arraycopy(payload, 0, crcInput, 1, payload.size)
            val expectedCrc = crc8(crcInput, 0, crcInput.size)
            if (expectedCrc != actualCrc) {
                dropFront(leaderStart + symbolSamples)
                continue
            }

            decoded.add(payload)
            dropFront(frameEnd)
        }

        val keep = max(symbolSamples * 16, minSamples * 2)
        if (rxBuffer.size > keep) {
            dropFront(rxBuffer.size - keep)
        }
        return decoded
    }

    private fun synthesizeSymbol(nibble: Int): FloatArray {
        val row = (nibble ushr 2) and 0x3
        val col = nibble and 0x3
        val out = FloatArray(symbolSamples)
        val rowRef = rowRefs[row]
        val colRef = colRefs[col]
        for (i in 0 until symbolSamples) {
            out[i] = rowRef[i] + colRef[i]
        }
        return out
    }

    private fun findLeaderStart(): Int? {
        val maxStart = rxBuffer.size - ((leaderSymbols + gapSymbols + syncNibbles.size) * symbolSamples)
        if (maxStart <= 0) {
            return null
        }
        var bestStart = -1
        var bestScore = Int.MIN_VALUE
        val coarseStep = max(8, symbolSamples / 6)
        var coarse = 0
        while (coarse <= maxStart) {
            val score = scoreLeaderAt(coarse)
            if (score > bestScore) {
                bestScore = score
                bestStart = coarse
            }
            coarse += coarseStep
        }
        if (bestStart < 0) {
            return null
        }
        val refineFrom = max(0, bestStart - coarseStep)
        val refineTo = min(maxStart, bestStart + coarseStep)
        val refineStep = max(1, coarseStep / 32)
        var candidate = refineFrom
        while (candidate <= refineTo) {
            val score = scoreLeaderAt(candidate)
            if (score > bestScore) {
                bestScore = score
                bestStart = candidate
            }
            candidate += refineStep
        }
        return if (bestScore >= leaderSymbols) bestStart else null
    }

    private fun scoreLeaderAt(start: Int): Int {
        var score = 0
        for (symbolIndex in 0 until leaderSymbols) {
            val windowStart = start + (symbolIndex * symbolSamples)
            if (windowRms(windowStart) < minLeaderRms) {
                return Int.MIN_VALUE
            }
            val nibble = decodeNibbleAtSample(windowStart) ?: return Int.MIN_VALUE
            if (nibble == 0) {
                score += 1
            }
        }
        return score
    }

    private fun refineSyncStart(nominalStart: Int): Int? {
        val maxStart = rxBuffer.size - (syncNibbles.size * symbolSamples)
        if (maxStart <= 0) {
            return null
        }
        val searchRadius = symbolSamples
        val from = max(0, nominalStart - searchRadius)
        val to = min(maxStart, nominalStart + searchRadius)
        var bestStart = -1
        var bestScore = Int.MIN_VALUE
        val coarseStep = max(4, symbolSamples / 24)
        var candidate = from
        while (candidate <= to) {
            val score = scoreSyncAt(candidate)
            if (score > bestScore) {
                bestScore = score
                bestStart = candidate
            }
            candidate += coarseStep
        }
        if (bestStart >= 0) {
            val fineFrom = max(from, bestStart - coarseStep)
            val fineTo = min(to, bestStart + coarseStep)
            for (fineCandidate in fineFrom..fineTo) {
                val score = scoreSyncAt(fineCandidate)
                if (score > bestScore) {
                    bestScore = score
                    bestStart = fineCandidate
                }
            }
        }
        return if (bestScore >= syncNibbles.size - 1) bestStart else null
    }

    private fun scoreSyncAt(start: Int): Int {
        var score = 0
        for (index in syncNibbles.indices) {
            val nibble = decodeNibbleAt(start, index) ?: return Int.MIN_VALUE
            if (nibble == syncNibbles[index]) {
                score += 1
            }
        }
        return score
    }

    private fun decodeNibbleAt(start: Int, nibbleIndex: Int): Int? {
        return decodeNibbleAtSample(start + (nibbleIndex * symbolSamples))
    }

    private fun decodeRepeatedNibbleAt(start: Int, nibbleIndex: Int): Int? {
        val values = IntArray(repetitionCount)
        for (offset in 0 until repetitionCount) {
            values[offset] = decodeNibbleAt(start, nibbleIndex + offset) ?: return null
        }
        return majorityNibble(values)
    }

    private fun decodeNibbleAtSample(windowStart: Int): Int? {
        val windowEnd = windowStart + symbolSamples
        if (windowStart < 0 || windowEnd > rxBuffer.size) {
            return null
        }

        val rowMagnitudes = FloatArray(4)
        val colMagnitudes = FloatArray(4)
        for (index in 0 until 4) {
            var rowI = 0f
            var rowQ = 0f
            var colI = 0f
            var colQ = 0f
            val rowRef = rowRefs[index]
            val colRef = colRefs[index]
            for (sampleIndex in 0 until symbolSamples) {
                val sample = rxBuffer[windowStart + sampleIndex]
                rowI += sample * rowRef[sampleIndex]
                rowQ += sample * rowRef[sampleIndex + symbolSamples]
                colI += sample * colRef[sampleIndex]
                colQ += sample * colRef[sampleIndex + symbolSamples]
            }
            rowMagnitudes[index] = sqrt((rowI * rowI) + (rowQ * rowQ))
            colMagnitudes[index] = sqrt((colI * colI) + (colQ * colQ))
        }
        val row = rowMagnitudes.indices.maxByOrNull { rowMagnitudes[it] } ?: return null
        val col = colMagnitudes.indices.maxByOrNull { colMagnitudes[it] } ?: return null
        return ((row shl 2) or col) and 0xF
    }

    private fun makeReference(freqHz: Int, gain: Float): FloatArray {
        val out = FloatArray(symbolSamples * 2)
        for (index in 0 until symbolSamples) {
            val phase = (2.0 * PI.toFloat() * freqHz.toFloat() * index.toFloat()) / sampleRateHz.toFloat()
            out[index] = gain * sin(phase.toDouble()).toFloat()
            out[index + symbolSamples] = gain * cos(phase.toDouble()).toFloat()
        }
        return out
    }

    private fun windowRms(start: Int): Float {
        val end = start + symbolSamples
        if (start < 0 || end > rxBuffer.size) {
            return 0f
        }
        var energy = 0f
        for (index in start until end) {
            val sample = rxBuffer[index]
            energy += sample * sample
        }
        return sqrt(energy / symbolSamples.toFloat())
    }

    private fun dropFront(count: Int) {
        if (count <= 0 || rxBuffer.isEmpty()) {
            return
        }
        rxBuffer.subList(0, minOf(count, rxBuffer.size)).clear()
    }

    private fun crc8(bytes: ByteArray, start: Int, length: Int): Int {
        var crc = 0
        for (index in 0 until length) {
            crc = crc xor (bytes[start + index].toInt() and 0xFF)
        }
        return crc and 0xFF
    }

    private fun majorityNibble(values: IntArray): Int? {
        if (values.isEmpty()) {
            return null
        }
        val counts = HashMap<Int, Int>()
        for (value in values) {
            counts[value] = (counts[value] ?: 0) + 1
        }
        return counts.maxWithOrNull(
            compareBy<Map.Entry<Int, Int>> { it.value }
                .thenByDescending { it.key }
        )?.key
    }
}

File: Sources/Cyrinx/DTMFCodec.swift
import Foundation

struct DTMFCodec {
    private let sampleRateHz: Int
    private let symbolSamples: Int
    private let txGainCap: Float
    private let leaderSymbols = 6
    private let gapSymbols = 2
    private let minLeaderRms: Float = 0.0025
    private let repetitionCount = 3
    private let syncNibbles: [UInt8] = [0x1, 0x2, 0x3, 0x4, 0xB, 0xC, 0xD, 0xE]
    private let rowHz = [697, 770, 852, 941]
    private let colHz = [1209, 1336, 1477, 1633]
    private let rowRefs: [(sin: [Float], cos: [Float])]
    private let colRefs: [(sin: [Float], cos: [Float])]
    private let symbolTable: [[Float]]
    private var rxBuffer: [Float] = []

    init(config: Config, symbolSamples: Int) {
        sampleRateHz = Int(min(max(config.sampleRateHz, 8_000), 192_000))
        self.symbolSamples = min(max(symbolSamples, 960), 8192)
        txGainCap = min(max(config.txGainCap, 0.01), 0.80)
        let symbolCount = self.symbolSamples
        let gain = txGainCap * 0.5
        var rows: [(sin: [Float], cos: [Float])] = []
        rows.reserveCapacity(rowHz.count)
        for freqHz in rowHz {
            rows.append(
                Self.makeReference(
                    freqHz: freqHz, sampleRateHz: sampleRateHz, symbolSamples: symbolCount, gain: gain))
        }
        rowRefs = rows
        var cols: [(sin: [Float], cos: [Float])] = []
        cols.reserveCapacity(colHz.count)
        for freqHz in colHz {
            cols.append(
                Self.makeReference(
                    freqHz: freqHz, sampleRateHz: sampleRateHz, symbolSamples: symbolCount, gain: gain))
        }
        colRefs = cols
        symbolTable = (0..<16).map { nibble in
            let row = max(0, min(3, nibble >> 2))
            let col = max(0, min(3, nibble & 0x3))
            var symbol = [Float](repeating: 0, count: symbolCount)
            for index in 0..<symbolCount {
                symbol[index] = rows[row].sin[index] + cols[col].sin[index]
            }
            return symbol
        }
    }

    mutating func encode(payload: Data) -> [Float] {
        let trimmed = Data(payload.prefix(255))
        var crcFrame = Data([UInt8(trimmed.count)])
        crcFrame.append(trimmed)
        let crc = UInt8(Self.crc8(crcFrame))

        var nibbles = syncNibbles
        let lenHi = UInt8((trimmed.count >> 4) & 0xF)
        let lenLo = UInt8(trimmed.count & 0xF)
        for _ in 0..<repetitionCount {
            nibbles.append(lenHi)
        }
        for _ in 0..<repetitionCount {
            nibbles.append(lenLo)
        }
        for byte in trimmed {
            let hi = UInt8((byte >> 4) & 0xF)
            let lo = UInt8(byte & 0xF)
            for _ in 0..<repetitionCount {
                nibbles.append(hi)
            }
            for _ in 0..<repetitionCount {
                nibbles.append(lo)
            }
        }
        let crcHi = UInt8((crc >> 4) & 0xF)
        let crcLo = UInt8(crc & 0xF)
        for _ in 0..<repetitionCount {
            nibbles.append(crcHi)
        }
        for _ in 0..<repetitionCount {
            nibbles.append(crcLo)
        }

        let totalSymbols = leaderSymbols + gapSymbols + nibbles.count
        var output = [Float](repeating: 0, count: totalSymbols * symbolSamples)
        var cursor = 0
        for _ in 0..<leaderSymbols {
            writeWaveform(symbolTable[0], into: &output, offset: cursor)
            cursor += symbolSamples
        }
        cursor += gapSymbols * symbolSamples
        for nibble in nibbles {
            writeWaveform(symbolTable[Int(nibble)], into: &output, offset: cursor)
            cursor += symbolSamples
        }
        return output
    }

    mutating func ingest(samples: UnsafeBufferPointer<Float>) -> [Data] {
        if samples.isEmpty {
            return []
        }
        rxBuffer.append(contentsOf: samples)
        return decodeAvailable()
    }

    mutating func decodeWaveform(_ samples: [Float]) -> [Data] {
        samples.withUnsafeBufferPointer { ingest(samples: $0) }
    }

    private mutating func decodeAvailable() -> [Data] {
        var decoded: [Data] = []
        let minSymbols = leaderSymbols + gapSymbols + syncNibbles.count + (4 * repetitionCount)
        let minSamples = minSymbols * symbolSamples

        while rxBuffer.count >= minSamples {
            guard let leaderStart = findLeaderStart() else {
                break
            }
            let nominalStart = leaderStart + ((leaderSymbols + gapSymbols) * symbolSamples)
            guard let start = refineSyncStart(near: nominalStart) else {
                dropFront(leaderStart + symbolSamples)
                continue
            }
            guard
                let lenHi = decodeRepeatedNibble(at: start, nibbleIndex: syncNibbles.count),
                let lenLo = decodeRepeatedNibble(at: start, nibbleIndex: syncNibbles.count + repetitionCount)
            else {
                dropFront(leaderStart + symbolSamples)
                continue
            }
            let length = Int((lenHi << 4) | lenLo)
            let totalNibbles =
                syncNibbles.count + (2 * repetitionCount) + (length * 2 * repetitionCount)
                + (2 * repetitionCount)
            let frameEnd = start + (totalNibbles * symbolSamples)
            if rxBuffer.count < frameEnd {
                if leaderStart > 0 {
                    dropFront(leaderStart)
                }
                break
            }

            var payload = Data(capacity: length)
            var nibbleIndex = syncNibbles.count + (2 * repetitionCount)
            var valid = true
            for _ in 0..<length {
                guard
                    let hi = decodeRepeatedNibble(at: start, nibbleIndex: nibbleIndex),
                    let lo = decodeRepeatedNibble(at: start, nibbleIndex: nibbleIndex + repetitionCount)
                else {
                    valid = false
                    break
                }
                payload.append((hi << 4) | lo)
                nibbleIndex += 2 * repetitionCount
            }
            guard valid else {
                dropFront(leaderStart + symbolSamples)
                continue
            }
            guard
                let crcHi = decodeRepeatedNibble(at: start, nibbleIndex: nibbleIndex),
                let crcLo = decodeRepeatedNibble(at: start, nibbleIndex: nibbleIndex + repetitionCount)
            else {
                dropFront(leaderStart + symbolSamples)
                continue
            }

            let actualCRC = Int((crcHi << 4) | crcLo)
            var crcFrame = Data([UInt8(length)])
            crcFrame.append(payload)
            let expectedCRC = Self.crc8(crcFrame)
            if expectedCRC != actualCRC {
                dropFront(leaderStart + symbolSamples)
                continue
            }

            decoded.append(payload)
            dropFront(frameEnd)
        }

        let keep = max(symbolSamples * 16, minSamples * 2)
        if rxBuffer.count > keep {
            dropFront(rxBuffer.count - keep)
        }
        return decoded
    }

    private func findLeaderStart() -> Int? {
        let maxStart = rxBuffer.count - ((leaderSymbols + gapSymbols + syncNibbles.count) * symbolSamples)
        if maxStart <= 0 {
            return nil
        }
        var bestStart = -1
        var bestScore = Int.min
        let coarseStep = max(8, symbolSamples / 6)
        var coarse = 0
        while coarse <= maxStart {
            let score = scoreLeader(at: coarse)
            if score > bestScore {
                bestScore = score
                bestStart = coarse
            }
            coarse += coarseStep
        }
        if bestStart < 0 {
            return nil
        }
        let refineFrom = max(0, bestStart - coarseStep)
        let refineTo = min(maxStart, bestStart + coarseStep)
        let refineStep = max(1, coarseStep / 32)
        var candidate = refineFrom
        while candidate <= refineTo {
            let score = scoreLeader(at: candidate)
            if score > bestScore {
                bestScore = score
                bestStart = candidate
            }
            candidate += refineStep
        }
        return bestScore >= leaderSymbols ? bestStart : nil
    }

    private func scoreLeader(at start: Int) -> Int {
        var score = 0
        for symbolIndex in 0..<leaderSymbols {
            let windowStart = start + (symbolIndex * symbolSamples)
            if windowRms(start: windowStart) < minLeaderRms {
                return Int.min
            }
            guard let nibble = decodeNibbleAtSample(windowStart) else {
                return Int.min
            }
            if nibble == 0 {
                score += 1
            }
        }
        return score
    }

    private func refineSyncStart(near nominalStart: Int) -> Int? {
        let maxStart = rxBuffer.count - (syncNibbles.count * symbolSamples)
        if maxStart <= 0 {
            return nil
        }
        let searchRadius = symbolSamples
        let from = max(0, nominalStart - searchRadius)
        let to = min(maxStart, nominalStart + searchRadius)
        var bestStart = -1
        var bestScore = Int.min
        let coarseStep = max(4, symbolSamples / 24)
        var candidate = from
        while candidate <= to {
            let score = scoreSync(at: candidate)
            if score > bestScore {
                bestScore = score
                bestStart = candidate
            }
            candidate += coarseStep
        }
        if bestStart >= 0 {
            let fineFrom = max(from, bestStart - coarseStep)
            let fineTo = min(to, bestStart + coarseStep)
            for candidate in fineFrom...fineTo {
                let score = scoreSync(at: candidate)
                if score > bestScore {
                    bestScore = score
                    bestStart = candidate
                }
            }
        }
        return bestScore >= syncNibbles.count - 1 ? bestStart : nil
    }

    private func scoreSync(at start: Int) -> Int {
        var score = 0
        for (index, expected) in syncNibbles.enumerated() {
            guard let nibble = decodeNibble(at: start, nibbleIndex: index) else {
                return Int.min
            }
            if nibble == expected {
                score += 1
            }
        }
        return score
    }

    private func decodeNibble(at start: Int, nibbleIndex: Int) -> UInt8? {
        decodeNibbleAtSample(start + (nibbleIndex * symbolSamples))
    }

    private func decodeRepeatedNibble(at start: Int, nibbleIndex: Int) -> UInt8? {
        var values = [UInt8]()
        values.reserveCapacity(repetitionCount)
        for offset in 0..<repetitionCount {
            guard let nibble = decodeNibble(at: start, nibbleIndex: nibbleIndex + offset) else {
                return nil
            }
            values.append(nibble)
        }
        return Self.majorityNibble(values)
    }

    private func decodeNibbleAtSample(_ windowStart: Int) -> UInt8? {
        let windowEnd = windowStart + symbolSamples
        if windowStart < 0 || windowEnd > rxBuffer.count {
            return nil
        }

        var rowMagnitudes = [Float](repeating: 0, count: 4)
        var colMagnitudes = [Float](repeating: 0, count: 4)
        for index in 0..<4 {
            var rowI: Float = 0
            var rowQ: Float = 0
            var colI: Float = 0
            var colQ: Float = 0
            for sampleIndex in 0..<symbolSamples {
                let sample = rxBuffer[windowStart + sampleIndex]
                rowI += sample * rowRefs[index].sin[sampleIndex]
                rowQ += sample * rowRefs[index].cos[sampleIndex]
                colI += sample * colRefs[index].sin[sampleIndex]
                colQ += sample * colRefs[index].cos[sampleIndex]
            }
            rowMagnitudes[index] = sqrt((rowI * rowI) + (rowQ * rowQ))
            colMagnitudes[index] = sqrt((colI * colI) + (colQ * colQ))
        }
        guard
            let row = rowMagnitudes.enumerated().max(by: { $0.element < $1.element })?.offset,
            let col = colMagnitudes.enumerated().max(by: { $0.element < $1.element })?.offset
        else {
            return nil
        }
        return UInt8(((row << 2) | col) & 0xF)
    }

    private mutating func dropFront(_ count: Int) {
        if count <= 0 || rxBuffer.isEmpty {
            return
        }
        rxBuffer.removeFirst(min(count, rxBuffer.count))
    }

    private func windowRms(start: Int) -> Float {
        let end = start + symbolSamples
        if start < 0 || end > rxBuffer.count {
            return 0
        }
        var energy: Float = 0
        for index in start..<end {
            let sample = rxBuffer[index]
            energy += sample * sample
        }
        return sqrt(energy / Float(symbolSamples))
    }

    private static func makeReference(
        freqHz: Int,
        sampleRateHz: Int,
        symbolSamples: Int,
        gain: Float
    ) -> (sin: [Float], cos: [Float]) {
        let values = (0..<symbolSamples).map { index -> (Float, Float) in
            let phase = (2.0 * Float.pi * Float(freqHz) * Float(index)) / Float(sampleRateHz)
            return (gain * Foundation.sin(phase), gain * Foundation.cos(phase))
        }
        return (values.map(\.0), values.map(\.1))
    }

    private static func crc8<S: Sequence>(_ bytes: S) -> Int where S.Element == UInt8 {
        var crc = 0
        for byte in bytes {
            crc ^= Int(byte)
        }
        return crc & 0xFF
    }

    private static func majorityNibble(_ values: [UInt8]) -> UInt8? {
        guard !values.isEmpty else {
            return nil
        }
        var counts = [UInt8: Int]()
        for value in values {
            counts[value, default: 0] += 1
        }
        return counts.max { lhs, rhs in
            if lhs.value == rhs.value {
                return lhs.key > rhs.key
            }
            return lhs.value < rhs.value
        }?.key
    }
}
